shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java

```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
     restartAlwaysRequired = isRestartAlwaysRequired();
   }
 
+  // Loosely based off of
+  // https://github.com/android/user-interface-samples/blob/master/WindowInsetsAnimation/app/src/main/java/com/google/android/samples/insetsanimation/RootViewDeferringInsetsCallback.kt
+  // When the IME is shown or hidden, it sends an onApplyWindowInsets call with the
+  // final state of the IME. This defers the final call to allow the animation to
+  // take place before re-calling onApplyWindowInsets after animation completion.
+  @VisibleForTesting
+  @TargetApi(30)
+  @RequiresApi(30)
+  @SuppressLint("NewApi")
+  class ImeSyncDeferringInsetsCallback extends WindowInsetsAnimation.Callback
+      implements View.OnApplyWindowInsetsListener {
+    private int overlayInsetTypes;
+    private int deferredInsetTypes;
+
+    private View view;
+    private WindowInsets lastWindowInsets;
+    private boolean deferredInsets = false;
+    private boolean started = false;
+
+    ImeSyncDeferringInsetsCallback(View view, int persistentInsetTypes, int deferredInsetTypes) {
+      super(WindowInsetsAnimation.Callback.DISPATCH_MODE_CONTINUE_ON_SUBTREE);
+      this.overlayInsetTypes = overlayInsetTypes;
+      this.deferredInsetTypes = deferredInsetTypes;
+      this.view = view;
+    }
+
+    @Override
+    public WindowInsets onApplyWindowInsets(View view, WindowInsets windowInsets) {
+      this.view = view;
+      if (!started) {
+        // Store the view and insets for us in onEnd() below
+        lastWindowInsets = windowInsets;
+      }
+      if (deferredInsets) {
+        return WindowInsets.CONSUMED;
+      }
+      return view.onApplyWindowInsets(windowInsets);
+    }
+
+    @TargetApi(30)
+    @Override
+    public void onPrepare(WindowInsetsAnimation animation) {
+      if ((animation.getTypeMask() & deferredInsetTypes) != 0) {
+        deferredInsets = true;
+      }
+    }
+
+    @TargetApi(30)
+    @Override
+    public WindowInsetsAnimation.Bounds onStart(
+        WindowInsetsAnimation animation, WindowInsetsAnimation.Bounds bounds) {
+      if (deferredInsets && (animation.getTypeMask() & deferredInsetTypes) != 0) {
```

Comment:
comments here too. What's the difference between onprepare and onstart? i.e. why not just set both deferredInsets and started to true here? Can you start and not prepare?

Reply:
I initially thought that both were necessary, turns out only start is required.